src/main/java/frc/robot/subsystems/color/SignalingSubsystem.java

```diff
@@ -0,0 +1,245 @@
+import frc.robot.subsystems.color.patterns.RainbowPattern;
+import frc.robot.subsystems.color.patterns.TransFlag;
+import java.util.ArrayList;
+import java.util.function.Consumer;
+
+public class SignalingSubsystem extends SubsystemBase {
+  private static final int patternUpdateFrequency = 10;
+
+  private final AddressableLED ledStrip;
+  private AddressableLEDBuffer ledBuffer;
+  // private final CANdle gridPositionCandle;
+
+  private static final int numberOfLEDS = Constants.LED_COUNT;
+
+  private int tick;
+  private int patternTick;
+  private Timer amplifierTimer;
+  private Timer rumbleTimer;
+  private double rumbleEndTime = 0;
+  private boolean isAllianceAmplified;
+  private boolean isOpponentAmplified;
+
+  private final boolean isRedAlliance;
+
+  private DisablePattern disablePattern = DisablePattern.getRandom();
+
+  private final Consumer<Double> driverRumbleConsumer;
+
+  public SignalingSubsystem(int ID, Consumer<Double> driverRumbleConsumer) {
+    this.driverRumbleConsumer = driverRumbleConsumer;
+
+    tick = 0;
+    patternTick = 0;
+    amplifierTimer = new Timer();
+    rumbleTimer = new Timer();
+    isAllianceAmplified = false;
+    isOpponentAmplified = false;
+
+    this.isRedAlliance = DriverStation.getAlliance().equals(DriverStation.Alliance.Red);
+
+    // Initialize LED Strip
+    ledStrip = new AddressableLED(ID);
+    ledBuffer = new AddressableLEDBuffer(numberOfLEDS);
+    ledStrip.setLength(numberOfLEDS);
+    ledStrip.setData(ledBuffer);
+    ledStrip.start();
+
+    // Setup Disable Patterns
+    TransFlag.numberOfLEDS = numberOfLEDS;
+    FireFlyPattern.numberOfLEDS = numberOfLEDS;
+    BIFlag.numberOfLEDS = numberOfLEDS;
+    RainbowPattern.numberOfLEDS = numberOfLEDS;
+  }
+
+  @Override
+  public void periodic() {
+    // Update Light Pattern
+    if (DriverStation.isDisabled()) updatePattern();
+
+    // End Signaling
+    if (rumbleTimer.get() > rumbleEndTime) {
+      rumbleTimer.reset();
+      driverRumbleConsumer.accept(0.0);
+      setFullStrip(RGB.BLACK);
+    }
+    // Alliance Amplification Timer
+    if (isAllianceAmplified) {
+      displayAmplificationTimer(
+          10 - (int) amplifierTimer.get(), isRedAlliance ? RGB.RED : RGB.BLUE);
+      if (amplifierTimer.get() > 10) {
+        isAllianceAmplified = false;
+        amplifierTimer.reset();
+        startSignal(Constants.AMPLIFICATION_RUMBLE_TIME, Constants.AMPLIFICATION_RUMBLE_INTENSITY);
+      }
+    }
+    // Opponent Amplification Timer
+    else if (isOpponentAmplified) {
+      displayAmplificationTimer(
+          10 - (int) amplifierTimer.get(), isRedAlliance ? RGB.BLUE : RGB.RED);
+      if (amplifierTimer.get() > 10) {
+        isOpponentAmplified = false;
+        amplifierTimer.reset();
+        startSignal(Constants.AMPLIFICATION_RUMBLE_TIME, Constants.AMPLIFICATION_RUMBLE_INTENSITY);
+      }
+    }
+  }
+
+  public void startAmplification(boolean isOpposingTeam) {
```

Comment:
All function parameters and variables should be declared `final` wherever possible.